# **3️⃣ Transfer Learning: What It Is & When to Use It 🤖🔄**

## **💡 Real-Life Analogy: A Footballer Switching Leagues ⚽**

Imagine **Cristiano Ronaldo** moves from **La Liga (Spain) to the Premier League (England)**.  
- He **doesn’t need to relearn football** from scratch.  
- He **applies his existing skills** but adapts to the new league’s **style & pace**.  
- He **fine-tunes specific aspects** (e.g., adjusting to Premier League’s physicality).

📌 **In machine learning, Transfer Learning works the same way! A model trained on one task can be reused for another, with minimal retraining.**

# **📌 What is Transfer Learning?**

✅ **Transfer Learning is a technique where a pre-trained model is adapted to solve a new but related problem.**  
✅ Instead of training a model **from scratch**, we **reuse knowledge** from a model trained on a **large dataset**.  
✅ It **saves time, improves accuracy, and requires less data** for the new task.

📌 **Mathematical Representation:**  
$$\text{New Model} = \text{Pre-Trained Model} + \text{Fine-Tuning on New Data}$$

✅ **Instead of starting from zero, we modify an existing model for our specific needs!**

# **📊 When Would You Use Transfer Learning?**

| **Scenario**                                   | **Why Use Transfer Learning?**                                                                 |
|------------------------------------------------|------------------------------------------------------------------------------------------------|
| **Image Classification (Football Players) ⚽📷**  | If a model is trained on **millions of general images**, we can adapt it for **identifying football players**. |
| **Sports Video Analysis 🎥🏀**                     | A model trained on **YouTube videos** can be fine-tuned for **detecting dribbles, shots, and passes**.           |
| **Medical Image Diagnosis 🏥🩻**                   | A model trained on **general images** can be adapted to **detect diseases in X-rays or MRIs**.                      |
| **Self-Driving Cars 🚗🛣️**                        | A model trained on **American roads** can be adapted for **UK roads** with minor adjustments.                      |
| **NLP (Chatbots, Sentiment Analysis) 🗣️📢**         | A language model trained on **Wikipedia data** can be fine-tuned for **customer support**.                         |

✅ **Transfer Learning is ideal when:**  
- **We have limited data** for the new task.  
- **Training a model from scratch is too expensive.**  
- **The new task is similar to the pre-trained model’s task.**

# **📊 Example 1: Transfer Learning for Football Player Recognition (Image Classification) ⚽📷**

📌 **Scenario:** You want to train a model to **recognize football players** from match photos.  
- Instead of training from scratch, we **use a pre-trained model like ResNet or VGG**.  
- Fine-tune it using **images of football players** to specialize the model.

📌 **Steps:**  
1️⃣ **Load a pre-trained model (ResNet50, trained on ImageNet)**.  
2️⃣ **Remove the last classification layer**.  
3️⃣ **Add a new layer for “Football Player Identification”**.  
4️⃣ **Fine-tune using our football images**.

✅ **Why Transfer Learning?**  
- We **reuse general image knowledge** from **millions of images** instead of training from scratch.  
- The model **only needs to learn football-specific details**.

# **📊 Example 2: Transfer Learning for NBA Video Analysis (Deep Learning) 🎥🏀**

📌 **Scenario:** You want to build a model that **identifies NBA plays (dunks, assists, 3-pointers, blocks) from video footage**.  
- Instead of training from scratch, we **use a model pre-trained on YouTube sports videos**.  
- Fine-tune it using **NBA-specific plays**.

📌 **Steps:**  
1️⃣ Use a pre-trained **ConvNet + LSTM model trained on general sports videos**.  
2️⃣ Fine-tune it using **NBA play-specific data**.  
3️⃣ **Freeze early layers** (to keep general motion recognition).  
4️⃣ **Train only the last few layers** (to specialize in NBA plays).

✅ **Why Transfer Learning?**  
- Video data is **expensive to label**, so using an existing model saves time.  
- The model **already understands general player movements** and just needs **NBA-specific tuning**.

# **📊 Example 3: Transfer Learning in NLP (Chatbots & Sentiment Analysis) 🗣️📢**

📌 **Scenario:** You want to build a **chatbot for football fans** that understands match discussions.  
- Instead of training from scratch, we **fine-tune GPT (or BERT) using football-related text data**.

📌 **Steps:**  
1️⃣ Start with a **pre-trained NLP model (GPT-4, BERT)**.  
2️⃣ Fine-tune using **football commentary, fan discussions, and match reports**.  
3️⃣ The model **adapts to football-specific conversations**.

✅ **Why Fine-Tune GPT?**  
- GPT **already knows general language**.  
- We **only need to adjust it for football conversations**.

# **📊 Advantages & Disadvantages of Transfer Learning**

| Feature            | Pros ✅                                                     | Cons ❌                              |
|--------------------|-------------------------------------------------------------|--------------------------------------|
| **Training Time**  | Saves **time & computation** by reusing models.             | Can still require **fine-tuning time**. |
| **Performance**    | Works well for **similar tasks**.                           | Can **fail if the new task is too different**. |
| **Data Requirements** | Needs **less data** for training.                        | Still requires **some domain-specific data**. |
| **Flexibility**    | Works in **images, video, NLP, speech**.                      | Some pre-trained models may not **adapt well**. |

✅ **Use Transfer Learning when:**  
- **You don’t have enough data.**  
- **Training from scratch is too expensive.**  
- **The new task is similar to the pre-trained model’s task.**

# **🔥 Final Thoughts**

1️⃣ **Fine-Tuning = Adjusting Pre-Trained Models for New, Related Tasks.**  
2️⃣ **It saves time & computational resources.**  
3️⃣ **Used in image recognition, video analysis, NLP, and self-driving sports analysis.**  
4️⃣ **Perfect for player detection, action classification, and chatbot creation.**

# **Deep Dive into Fine-Tuning Pre-Trained Models for Sports Analytics ⚽🏀🤖**

## **💡 Real-Life Analogy: Adapting a Star Player to a New League 🌍⚽**

Imagine you’re **a coach signing Lionel Messi** for a new club.  
- **Pre-Trained Model:** Messi already knows how to **dribble, shoot, and pass** (general skills).  
- **Fine-Tuning:** You **adjust his playstyle** to **fit the new team’s tactics** (specific adaptation).  
- **Frozen Layers:** His **fundamental skills don’t change**—only minor tweaks are made.

📌 **Fine-tuning a deep learning model works the same way! We adapt a pre-trained model for a new but related task.**

# **📌 What is Fine-Tuning in Transfer Learning?**

✅ **Fine-tuning = Taking a pre-trained model and training it further on new, specific data.**  
✅ **Steps:**  
1️⃣ **Use a pre-trained model** (trained on a large dataset like ImageNet for images, YouTube for videos, or GPT for text).  
2️⃣ **Freeze some layers** (retain general knowledge, like Messi’s skills).  
3️⃣ **Replace the last few layers** (adapt to the new sport or dataset).  
4️⃣ **Train only the new layers** on domain-specific data (fine-tuning for NBA, EPL, etc.).

📌 **Key Benefit:** Faster training with **higher accuracy on the new task**.

# **📊 Example 1: Fine-Tuning a Model for Football Player Recognition (Computer Vision) 📷⚽**

📌 **Scenario:** You want to **identify football players** in match photos.  
- **Problem:** Training from scratch requires **millions of images**.  
- **Solution:** Use **ResNet50 (pre-trained on ImageNet) and fine-tune it** for football players.

📌 **How?**  
1️⃣ **Load ResNet50 (Pre-Trained on ImageNet).**  
2️⃣ **Freeze Early Layers** → Keep general image features (edges, textures, patterns).  
3️⃣ **Replace Final Layers** → Train on football player images.  
4️⃣ **Fine-Tune** → Improve classification for football-specific features.

📌 **Python Implementation (Fine-Tuning ResNet for Football Player Recognition)**

📌 **Python Implementation (Using Pre-Trained ResNet50 for Football Players)**

### 1. Importing Required Libraries for Image Transfer Learning

`ResNet50` is a popular pre-trained model for image classification.  
`Model` is used to create a new model architecture.  
`Dense` is a fully connected layer for classification.  
`Flatten` is used to convert 2D feature maps to 1D.  

In [1]:
from tensorflow.keras.applications import ResNet50
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Dense, Flatten

### 2. Load Pre-Trained ResNet50 and Remove the Final Layer`


`weights="imagenet"` loads the model trained on ImageNet dataset.  
`include_top=False` removes the final classification layer.  
`input_shape=(224, 224, 3)` specifies the input image dimensions, typically 224x224 pixelx and 3 channels (RGB).  

In [2]:
# Load Pre-Trained Model (without the final classification layer)
base_model = ResNet50(weights="imagenet", include_top=False, input_shape=(224, 224, 3))
base_model.summary()  # Optional: view the model architecture

2025-02-10 15:09:24.987944: I metal_plugin/src/device/metal_device.cc:1154] Metal device set to: Apple M4 Pro
2025-02-10 15:09:24.987972: I metal_plugin/src/device/metal_device.cc:296] systemMemory: 24.00 GB
2025-02-10 15:09:24.987978: I metal_plugin/src/device/metal_device.cc:313] maxCacheSize: 8.00 GB
I0000 00:00:1739200164.988334 17254431 pluggable_device_factory.cc:305] Could not identify NUMA node of platform GPU ID 0, defaulting to 0. Your kernel may not have been built with NUMA support.
I0000 00:00:1739200164.988591 17254431 pluggable_device_factory.cc:271] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 0 MB memory) -> physical PluggableDevice (device: 0, name: METAL, pci bus id: <undefined>)


Model: "resnet50"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer         │ (None, 224, 224,  │          0 │ -                 │
│ (InputLayer)        │ 3)                │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv1_pad           │ (None, 230, 230,  │          0 │ input_layer[0][0] │
│ (ZeroPadding2D)     │ 3)                │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv1_conv (Conv2D) │ (None, 112, 112,  │      9,472 │ conv1_pad[0][0]   │
│                     │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv1_bn            │ (None, 112, 112,  │        256 │ conv1_conv[0][0]  │
│ (BatchNormalizatio… │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv1_relu          │ (None, 112, 112,  │          0 │ conv1_bn[0][0]    │
│ (Activation)        │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ pool1_pad           │ (None, 114, 114,  │          0 │ conv1_relu[0][0]  │
│ (ZeroPadding2D)     │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ pool1_pool          │ (None, 56, 56,    │          0 │ pool1_pad[0][0]   │
│ (MaxPooling2D)      │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_1_conv │ (None, 56, 56,    │      4,160 │ pool1_pool[0][0]  │
│ (Conv2D)            │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_1_bn   │ (None, 56, 56,    │        256 │ conv2_block1_1_c… │
│ (BatchNormalizatio… │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_1_relu │ (None, 56, 56,    │          0 │ conv2_block1_1_b… │
│ (Activation)        │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_2_conv │ (None, 56, 56,    │     36,928 │ conv2_block1_1_r… │
│ (Conv2D)            │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_2_bn   │ (None, 56, 56,    │        256 │ conv2_block1_2_c… │
│ (BatchNormalizatio… │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_2_relu │ (None, 56, 56,    │          0 │ conv2_block1_2_b… │
│ (Activation)        │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_0_conv │ (None, 56, 56,    │     16,640 │ pool1_pool[0][0]  │
│ (Conv2D)            │ 256)              │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_3_conv │ (None, 56, 56,    │     16,640 │ conv2_block1_2_r… │
│ (Conv2D)            │ 256)              │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_0_bn   │ (None, 56, 56,    │      1,024 │ conv2_block1_0_c… │
│ (BatchNormalizatio… │ 256)              │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_3_bn   │ (None, 56, 56,    │      1,024 │ conv2_block1_3_c

 Total params: 23,587,712 (89.98 MB)

 Trainable params: 23,534,592 (89.78 MB)

 Non-trainable params: 53,120 (207.50 KB)

`Total params` are the model's total trainable parameters.  
`Trainable params` are the parameters that can be updated during training.  
`Non-trainable params` are the pre-trained parameters that remain fixed.  

### 3. Add Custom Layers and Create the New Model


`Dense` is a fully connected layer with 512 neurons and ReLU activation.
`Dense` is the final classification layer with 10 neurons (for 10 football players) and softmax activation.


In [3]:
# Add custom layers on top of the base model
x = Flatten()(base_model.output)
x = Dense(512, activation="relu")(x)
output_layer = Dense(10, activation="softmax")(x)  # 10 classes for football players

# Create the new model
model = Model(inputs=base_model.input, outputs=output_layer)
model.summary()  # Optional: inspect the complete model architecture

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer         │ (None, 224, 224,  │          0 │ -                 │
│ (InputLayer)        │ 3)                │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv1_pad           │ (None, 230, 230,  │          0 │ input_layer[0][0] │
│ (ZeroPadding2D)     │ 3)                │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv1_conv (Conv2D) │ (None, 112, 112,  │      9,472 │ conv1_pad[0][0]   │
│                     │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv1_bn            │ (None, 112, 112,  │        256 │ conv1_conv[0][0]  │
│ (BatchNormalizatio… │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv1_relu          │ (None, 112, 112,  │          0 │ conv1_bn[0][0]    │
│ (Activation)        │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ pool1_pad           │ (None, 114, 114,  │          0 │ conv1_relu[0][0]  │
│ (ZeroPadding2D)     │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ pool1_pool          │ (None, 56, 56,    │          0 │ pool1_pad[0][0]   │
│ (MaxPooling2D)      │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_1_conv │ (None, 56, 56,    │      4,160 │ pool1_pool[0][0]  │
│ (Conv2D)            │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_1_bn   │ (None, 56, 56,    │        256 │ conv2_block1_1_c… │
│ (BatchNormalizatio… │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_1_relu │ (None, 56, 56,    │          0 │ conv2_block1_1_b… │
│ (Activation)        │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_2_conv │ (None, 56, 56,    │     36,928 │ conv2_block1_1_r… │
│ (Conv2D)            │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_2_bn   │ (None, 56, 56,    │        256 │ conv2_block1_2_c… │
│ (BatchNormalizatio… │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_2_relu │ (None, 56, 56,    │          0 │ conv2_block1_2_b… │
│ (Activation)        │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_0_conv │ (None, 56, 56,    │     16,640 │ pool1_pool[0][0]  │
│ (Conv2D)            │ 256)              │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_3_conv │ (None, 56, 56,    │     16,640 │ conv2_block1_2_r… │
│ (Conv2D)            │ 256)              │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_0_bn   │ (None, 56, 56,    │      1,024 │ conv2_block1_0_c… │
│ (BatchNormalizatio… │ 256)              │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2_block1_3_bn   │ (None, 56, 56,    │      1,024 │ conv2_block1_3_c

 Total params: 74,973,578 (286.00 MB)

 Trainable params: 74,920,458 (285.80 MB)

 Non-trainable params: 53,120 (207.50 KB)

### 4. Freeze Pre-Trained Layers and Compile the Model



We freeze the pre-trained `ResNet50` layers to retain their knowledge and prevent overfitting.
The model is compiled with the `adam` optimizer and `categorical_crossentropy` loss function.

In [4]:
# Freeze all layers in the base model so their weights are not updated during training
for layer in base_model.layers:
    layer.trainable = False

# Compile the model with a chosen optimizer and loss function
model.compile(optimizer="adam", loss="categorical_crossentropy", metrics=["accuracy"])

In [5]:
from tensorflow.keras.applications import ResNet50
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.optimizers import Adam

# Load Pre-Trained Model (Without Final Layer)
base_model = ResNet50(weights="imagenet", include_top=False, input_shape=(224, 224, 3))

# Freeze Early Layers (Keep General Features)
for layer in base_model.layers[:-10]:
    layer.trainable = False

# Add Custom Layers for Football Players
x = Flatten()(base_model.output)
x = Dense(512, activation="relu")(x)
output_layer = Dense(10, activation="softmax")(x)  # 10 football players

# Create New Model
model = Model(inputs=base_model.input, outputs=output_layer)

# Compile Model
model.compile(optimizer=Adam(learning_rate=0.0001), loss="categorical_crossentropy", metrics=["accuracy"])

✅ **Why Fine-Tune ResNet?**  
- We **reuse general image knowledge** (faces, jerseys, backgrounds).  
- The model **only learns football-specific player features**.  
- **Faster training** (only last few layers are updated).

# **📊 Example 2: Fine-Tuning for NBA Play Recognition (Video Analysis) 🎥🏀**

📌 **Scenario:** You want to **identify key NBA plays** (dunks, assists, 3-pointers, blocks) from game videos.  
- **Problem:** Training from scratch on videos **requires massive datasets & GPUs**.  
- **Solution:** Use a **pre-trained ConvNet+LSTM model trained on general sports videos** and fine-tune it for NBA.

📌 **How?**  
1️⃣ Use a pre-trained **3D CNN (e.g., I3D or C3D) trained on general sports videos.**  
2️⃣ **Freeze Early Layers** → Keep motion features (player movement, ball motion).  
3️⃣ **Replace Final Layers** → Train only on NBA-specific play clips.  
4️⃣ **Fine-Tune** → Improve recognition of NBA-specific movements.

📌 **Python Implementation (Fine-Tuning a Pre-Trained Video Model)**

## Why Use MobileNetV2 for Transfer Learning in NBA Video Analysis?



`MobileNetV2` is a lightweight and efficient convolutional neural network designed primarily for image classification. It has several advantages:

- **Efficiency:** Its smaller footprint makes it ideal for edge devices and mobile applications.
- **Pre-trained Weights:** `MobileNetV2` is available with pre-trained weights (e.g. on `ImageNet`), so you can leverage learned features without training from scratch.
- **Flexibility:** You can easily freeze most of its layers and add custom layers on top for a specialized task—such as classifying NBA plays (dunks, 3-pointers, blocks, etc.).

In the initial demonstration, I simulated a generic ConvNet+LSTM model to illustrate the transfer learning workflow. However, if you already have code that uses `MobileNetV2`, it’s an excellent alternative. Below is the original code (with slight modifications for clarity) that loads `MobileNetV2`, freezes most of its layers, and adds custom dense layers for NBA play classification.


### 1. Import Required Libraries



Here, we import `TensorFlow` and `Keras` modules needed to load `MobileNetV2` and to build our custom model.


In [6]:
import tensorflow as tf
from tensorflow.keras.applications import MobileNetV2
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam


### 2. Load Pre-Trained MobileNetV2 and Freeze Early Layers


We load `MobileNetV2` with `ImageNet` weights and exclude its top classification layer. Then we freeze most layers (all except the last 5) so that the network retains the general feature extraction capabilities.


In [7]:
# Load Pre-Trained MobileNetV2 (for processing video frames)
base_model = MobileNetV2(weights="imagenet", include_top=False, input_shape=(224, 224, 3))

# Freeze all layers except the last 5 to retain general features while allowing some fine-tuning
for layer in base_model.layers[:-5]:
    layer.trainable = False

# Optional: Print out trainable status for verification
for i, layer in enumerate(base_model.layers):
    print(f"Layer {i}: {layer.name} - Trainable: {layer.trainable}")

# View the MobileNetV2 architecture summary
base_model.summary()


Layer 0: input_layer_2 - Trainable: False
Layer 1: Conv1 - Trainable: False
Layer 2: bn_Conv1 - Trainable: False
Layer 3: Conv1_relu - Trainable: False
Layer 4: expanded_conv_depthwise - Trainable: False
Layer 5: expanded_conv_depthwise_BN - Trainable: False
Layer 6: expanded_conv_depthwise_relu - Trainable: False
Layer 7: expanded_conv_project - Trainable: False
Layer 8: expanded_conv_project_BN - Trainable: False
Layer 9: block_1_expand - Trainable: False
Layer 10: block_1_expand_BN - Trainable: False
Layer 11: block_1_expand_relu - Trainable: False
Layer 12: block_1_pad - Trainable: False
Layer 13: block_1_depthwise - Trainable: False
Layer 14: block_1_depthwise_BN - Trainable: False
Layer 15: block_1_depthwise_relu - Trainable: False
Layer 16: block_1_project - Trainable: False
Layer 17: block_1_project_BN - Trainable: False
Layer 18: block_2_expand - Trainable: False
Layer 19: block_2_expand_BN - Trainable: False
Layer 20: block_2_expand_relu - Trainable: False
Layer 21: block_2_d

Model: "mobilenetv2_1.00_224"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer_2       │ (None, 224, 224,  │          0 │ -                 │
│ (InputLayer)        │ 3)                │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Conv1 (Conv2D)      │ (None, 112, 112,  │        864 │ input_layer_2[0]… │
│                     │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ bn_Conv1            │ (None, 112, 112,  │        128 │ Conv1[0][0]       │
│ (BatchNormalizatio… │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Conv1_relu (ReLU)   │ (None, 112, 112,  │          0 │ bn_Conv1[0][0]    │
│                     │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_dept… │ (None, 112, 112,  │        288 │ Conv1_relu[0][0]  │
│ (DepthwiseConv2D)   │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_dept… │ (None, 112, 112,  │        128 │ expanded_conv_de… │
│ (BatchNormalizatio… │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_dept… │ (None, 112, 112,  │          0 │ expanded_conv_de… │
│ (ReLU)              │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_proj… │ (None, 112, 112,  │        512 │ expanded_conv_de… │
│ (Conv2D)            │ 16)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_proj… │ (None, 112, 112,  │         64 │ expanded_conv_pr… │
│ (BatchNormalizatio… │ 16)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_expand      │ (None, 112, 112,  │      1,536 │ expanded_conv_pr… │
│ (Conv2D)            │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_expand_BN   │ (None, 112, 112,  │        384 │ block_1_expand[0… │
│ (BatchNormalizatio… │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_expand_relu │ (None, 112, 112,  │          0 │ block_1_expand_B… │
│ (ReLU)              │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_pad         │ (None, 113, 113,  │          0 │ block_1_expand_r… │
│ (ZeroPadding2D)     │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_depthwise   │ (None, 56, 56,    │        864 │ block_1_pad[0][0] │
│ (DepthwiseConv2D)   │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_depthwise_… │ (None, 56, 56,    │        384 │ block_1_depthwis… │
│ (BatchNormalizatio… │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_depthwise_… │ (None, 56, 56,    │          0 │ block_1_depthwis… │
│ (ReLU)              │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_project     │ (None, 56, 56,    │      2,304 │ block_1_depthwis

 Total params: 2,257,984 (8.61 MB)

 Trainable params: 720,000 (2.75 MB)

 Non-trainable params: 1,537,984 (5.87 MB)

### 3. Add Custom Layers for NBA Play Classification


We now add custom layers on top of `MobileNetV2`. The idea is to fine-tune the network to identify NBA-specific plays (e.g., dunks, 3-pointers, blocks, etc.). In this example, we assume there are 4 categories. We add several Dense layers after flattening the feature maps.


In [8]:
# Add custom layers on top of the base MobileNetV2 model
x = Flatten()(base_model.output)
x = Dense(512, activation="relu")(x)
x = Dense(128, activation="relu")(x)
output_layer = Dense(4, activation="softmax")(x)  # 4 NBA play categories

# Create the new model
model = Model(inputs=base_model.input, outputs=output_layer)

# View the new model's architecture
model.summary()


Model: "functional_2"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer_2       │ (None, 224, 224,  │          0 │ -                 │
│ (InputLayer)        │ 3)                │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Conv1 (Conv2D)      │ (None, 112, 112,  │        864 │ input_layer_2[0]… │
│                     │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ bn_Conv1            │ (None, 112, 112,  │        128 │ Conv1[0][0]       │
│ (BatchNormalizatio… │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Conv1_relu (ReLU)   │ (None, 112, 112,  │          0 │ bn_Conv1[0][0]    │
│                     │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_dept… │ (None, 112, 112,  │        288 │ Conv1_relu[0][0]  │
│ (DepthwiseConv2D)   │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_dept… │ (None, 112, 112,  │        128 │ expanded_conv_de… │
│ (BatchNormalizatio… │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_dept… │ (None, 112, 112,  │          0 │ expanded_conv_de… │
│ (ReLU)              │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_proj… │ (None, 112, 112,  │        512 │ expanded_conv_de… │
│ (Conv2D)            │ 16)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_proj… │ (None, 112, 112,  │         64 │ expanded_conv_pr… │
│ (BatchNormalizatio… │ 16)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_expand      │ (None, 112, 112,  │      1,536 │ expanded_conv_pr… │
│ (Conv2D)            │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_expand_BN   │ (None, 112, 112,  │        384 │ block_1_expand[0… │
│ (BatchNormalizatio… │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_expand_relu │ (None, 112, 112,  │          0 │ block_1_expand_B… │
│ (ReLU)              │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_pad         │ (None, 113, 113,  │          0 │ block_1_expand_r… │
│ (ZeroPadding2D)     │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_depthwise   │ (None, 56, 56,    │        864 │ block_1_pad[0][0] │
│ (DepthwiseConv2D)   │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_depthwise_… │ (None, 56, 56,    │        384 │ block_1_depthwis… │
│ (BatchNormalizatio… │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_depthwise_… │ (None, 56, 56,    │          0 │ block_1_depthwis… │
│ (ReLU)              │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_project     │ (None, 56, 56,    │      2,304 │ block_1_depthwis

 Total params: 34,437,316 (131.37 MB)

 Trainable params: 32,899,332 (125.50 MB)

 Non-trainable params: 1,537,984 (5.87 MB)

### 4. Compile the Model


We compile our new NBA transfer model using the `Adam` optimizer with a low learning rate (`0.0001`) and categorical crossentropy loss. This configuration is standard for fine-tuning classification models.


In [9]:
model.compile(optimizer=Adam(learning_rate=0.0001),
              loss="categorical_crossentropy",
              metrics=["accuracy"])


## Dummy ConvNet+LSTM Model for Transfer Learning


### Model Components:


- **Input Representation:**  
  - The model accepts a video as a sequence of `10` frames, each of size `224×224×3`.
  
- **Frame-Level Feature Extraction:**  
  - A *TimeDistributed* wrapper applies a small convolutional neural network (`CNN`) to each frame.  
  - This CNN (referred to as `conv_base`) extracts spatial features from each frame, outputting a 64-dimensional feature vector per frame.
  
- **Temporal Dynamics Modeling:**  
  - An `LSTM` layer processes the sequence of frame features to capture temporal (motion) information across the video.
  - The `LSTM` outputs a `128-dimensional` feature vector summarizing the video.
  
- **Pre-Trained Head:**  
  - A Dense layer (named `pretrained_output`) maps the `LSTM` output to 5 general sports classes.
  - This head simulates a model that was originally trained on a broad sports dataset.


### Transfer Learning Strategy:


1. **Freezing the Base Layers:**  
   - The early layers (the `CNN` inside the TimeDistributed layer and the `LSTM` layer) are frozen.  
   - This ensures that the general visual and motion features learned during pre-training remain unchanged.

2. **Replacing the Final Layer:**  
   - The pre-trained `Dense` output layer is removed or bypassed.
   - New custom `Dense` layers are added to specialize the model for NBA play classification (e.g., dunks, 3-pointers, blocks).
   - The new head outputs predictions for the target number of classes (in our example, 3 classes).



### Pros and Cons of This Approach:


- **Pros:**
  - **Educational Value:** Clearly illustrates the transfer learning process by building a simplified, modular model.
  - **Concept Demonstration:** Shows how freezing early layers and fine-tuning only the final layers can adapt a general model to a specific task.
  
- **Cons:**
  - **Simplified Architecture:** The dummy model is a simplified simulation and may not perform as well as state‑of‑the‑art networks.
  - **Production Limitations:** It is intended mainly for demonstration and educational purposes rather than high-accuracy, production‑level applications.
  


### Why Switch to MobileNetV2?


`MobileNetV2` is a proven, efficient architecture that:
- **Offers Better Performance:** With pre-trained weights on `ImageNet`, it provides a robust feature extractor.
- **Is Resource-Efficient:** Its lightweight design makes it ideal for applications on devices with limited resources (e.g., mobile devices or edge computing).
- **Simplifies the Workflow:** The pre-trained `MobileNetV2` can be easily integrated and fine-tuned for NBA-specific tasks, often achieving better accuracy in real-world applications.

In the following cells, we adopt the `MobileNetV2`-based approach to build our NBA video analysis model.


### 1. Import Required Libraries for Video Transfer Learning

In this cell, we import `TensorFlow` and `Keras` modules necessary to build and modify our `ConvNet`+`LSTM` model.

In [10]:
import tensorflow as tf
from tensorflow.keras.models import Model
from tensorflow.keras.layers import (Input, Conv2D, MaxPooling2D, Flatten, 
                                     Dense, TimeDistributed, LSTM, GlobalAveragePooling2D)

### 2. Build a Dummy Pre-Trained ConvNet+LSTM Model



Since we don't have an actual pre-trained model file, we simulate one. This dummy model processes a sequence of `10` frames (each `224×224×3`) through a small `CNN` followed by an `LSTM` layer, outputting predictions for 5 sports classes.

In [11]:
# Define input shape for a video: (num_frames, height, width, channels)
num_frames = 10
frame_height, frame_width, channels = 224, 224, 3

# Input layer for video data
video_input = Input(shape=(num_frames, frame_height, frame_width, channels), name='video_input')

# Process each frame using the same ConvNet (using TimeDistributed)
conv_base = tf.keras.Sequential([
    Conv2D(32, (3,3), activation='relu', padding='same'),
    MaxPooling2D((2,2)),
    Conv2D(64, (3,3), activation='relu', padding='same'),
    GlobalAveragePooling2D()
], name='conv_base')

# Apply the conv_base to each frame
time_distributed = TimeDistributed(conv_base, name='frame_feature_extractor')(video_input)

# Use an LSTM to model temporal dynamics
lstm_out = LSTM(128, return_sequences=False, name='lstm_layer')(time_distributed)

# Add a dummy classification layer for the pre-trained model
pretrained_output = Dense(5, activation='softmax', name='pretrained_output')(lstm_out)

# Create the pre-trained model
pretrained_model = Model(inputs=video_input, outputs=pretrained_output, name='pretrained_convnet_lstm')
pretrained_model.summary()

Model: "pretrained_convnet_lstm"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ video_input (InputLayer)        │ (None, 10, 224, 224,   │             0 │
│                                 │ 3)                     │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ frame_feature_extractor         │ (None, 10, 64)         │        19,392 │
│ (TimeDistributed)               │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm_layer (LSTM)               │ (None, 128)            │        98,816 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ pretrained_output (Dense)       │ (None, 5)              │           645 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 118,853 (464.27 KB)

 Trainable params: 118,853 (464.27 KB)

 Non-trainable params: 0 (0.00 B)

### 3. Freeze Early Layers of the Pre-Trained Video Model



This cell freezes the `CNN` and `LSTM` layers to preserve the general features learned during pre-training.

In [12]:
# Freeze the conv_base and LSTM layers (i.e. all layers except the last Dense layer)
for layer in pretrained_model.layers:
    if layer.name in ['conv_base', 'frame_feature_extractor', 'lstm_layer']:
        layer.trainable = False

# Verify trainability of layers
for i, layer in enumerate(pretrained_model.layers):
    print(f"Layer {i}: {layer.name} - Trainable: {layer.trainable}")

Layer 0: video_input - Trainable: True
Layer 1: frame_feature_extractor - Trainable: False
Layer 2: lstm_layer - Trainable: False
Layer 3: pretrained_output - Trainable: True


### 4. Replace Final Classification Layers with NBA-Specific Layers



We remove the original `Dense` layer and add new `Dense` layers to adapt the model for NBA play classification (3 classes).

In [13]:
# Remove the pre-trained classification output and get the output from the LSTM layer
lstm_features = pretrained_model.get_layer('lstm_layer').output

# Add new dense layers for NBA play classification (3 classes: dunks, 3-pointers, blocks)
x = Dense(256, activation='relu', name='nba_dense')(lstm_features)
nba_output = Dense(3, activation='softmax', name='nba_output')(x)

# Create a new model for NBA play recognition using the pre-trained base
nba_model = Model(inputs=pretrained_model.input, outputs=nba_output, name='nba_transfer_model')
nba_model.summary()

Model: "nba_transfer_model"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ video_input (InputLayer)        │ (None, 10, 224, 224,   │             0 │
│                                 │ 3)                     │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ frame_feature_extractor         │ (None, 10, 64)         │        19,392 │
│ (TimeDistributed)               │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm_layer (LSTM)               │ (None, 128)            │        98,816 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ nba_dense (Dense)               │ (None, 256)            │        33,024 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ nba_output (Dense)              │ (None, 3)              │           771 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 152,003 (593.76 KB)

 Trainable params: 33,795 (132.01 KB)

 Non-trainable params: 118,208 (461.75 KB)

### 5. Compile the NBA Transfer Model



Compile the NBA model using `Adam` optimizer and categorical cross-entropy loss, appropriate for multi-class classification.

In [14]:
nba_model.compile(optimizer='adam', 
                  loss='categorical_crossentropy', 
                  metrics=['accuracy'])

### Explanation of Model Summaries



1. **Pre-Trained Model Summary:**  
   - **Input Layer:** Accepts a video sequence of 10 frames, each of size 224×224×3.  
   - **Frame Feature Extractor:** A TimeDistributed layer applies a small CNN (our `conv_base`) to each frame, outputting a 64-dimensional vector per frame.  
   - **LSTM Layer:** Processes the sequence of features and outputs a 128-dimensional vector, capturing temporal dynamics.  
   - **Pretrained Output:** A Dense layer producing outputs for 5 general sports classes.  
   - **Trainable Status:** The base layers are frozen (non-trainable), while the final Dense layer is trainable.

2. **NBA Transfer Model Summary:**  
   - **Reuse of Base Layers:** The same video input, CNN, and LSTM layers are used as in the pre-trained model (and remain frozen).  
   - **New Custom Layers:**  
     - `nba_dense`: A Dense layer with 256 units that learns NBA-specific features from the LSTM output.  
     - `nba_output`: A final Dense layer with 3 units (one for each NBA play class) using softmax activation for classification.  
   - **Trainable Parameters:** Only the new NBA-specific layers are trainable while the pre-trained base remains unchanged.

This structure leverages transfer learning by reusing general motion and spatial features from a broader sports video dataset and specializing them for NBA play recognition.

✅ **Why Fine-Tune a Video Model?**  
- It **already understands player movement** from previous training.  
- We **only adjust it for NBA-specific plays**.  
- **Saves weeks of training time** compared to starting from scratch.

# **📊 Example 3: Transfer Learning in NLP (Chatbots & Sentiment Analysis) 🗣️📢**

📌 **Scenario:** You want to build a **chatbot for football fans** that understands match discussions.  
- Instead of training from scratch, we **fine-tune GPT (or BERT) using football-related text data**.

📌 **Steps:**  
1️⃣ Start with a **pre-trained NLP model (GPT-4, BERT)**.  
2️⃣ Fine-tune using **football commentary, fan discussions, and match reports**.  
3️⃣ The model **adapts to football-specific conversations**.

📌 **Python Implementation (Fine-Tuning GPT for Football Chatbot)**

### 1. Import Required Libraries


We first import the necessary modules from the Hugging Face Transformers library. These include:
- `TFGPT2LMHeadModel`: The pre-trained GPT-2 model for language modeling.
- `GPT2Tokenizer`: The tokenizer for processing text inputs.
- `TFTrainingArguments`: The training arguments for fine-tuning the model.

In [21]:
from transformers import TFGPT2LMHeadModel, GPT2Tokenizer, TFTrainingArguments
import tensorflow as tf

print("Libraries imported successfully.")

Libraries imported successfully.


### 2. Load Pre-Trained GPT-2 Model and Tokenizer



We load the TensorFlow version of `GPT‑2` (`TFGPT2LMHeadModel`) along with its `GPT‑2 tokenizer`.  
If the tokenizer does not have a pad token, we set it to the end‑of‑sequence token (`eos_token`).


In [ ]:
# Load the TensorFlow version of the pre-trained GPT‑2 model and tokenizer from Hugging Face
model = TFGPT2LMHeadModel.from_pretrained("gpt2")
tokenizer = GPT2Tokenizer.from_pretrained("gpt2")

# Ensure the tokenizer has a pad token; if not, set it to the eos_token
if tokenizer.pad_token is None:
    tokenizer.pad_token = tokenizer.eos_token

print("Model and tokenizer loaded successfully.")

All PyTorch model weights were used when initializing TFGPT2LMHeadModel.

All the weights of TFGPT2LMHeadModel were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFGPT2LMHeadModel for predictions without further training.


Model and tokenizer loaded successfully.


### 3. Prepare and Tokenize Football-Specific Training Data

We define a small dataset of football-related commentary and tokenize it using the GPT‑2 tokenizer.  
We specify `truncation=True`, `padding=True`, and a fixed `max_length` (here, 64 tokens) so that each example is uniform.

In [28]:
# Define football-specific training data (list of strings)
train_data = [
    "Liverpool dominated possession but lacked clinical finishing.", 
    "Messi’s dribbling was unstoppable against Bayern’s high press."
]

# Tokenize the data using the GPT2 tokenizer.
# Using return_tensors="tf" ensures the output is a TensorFlow tensor.
train_encodings = tokenizer(train_data, truncation=True, padding=True, max_length=64, return_tensors="tf")

# Display the tokenized output for verification (optional)
print("Tokenized training data:")
print(train_encodings)


Tokenized training data:
{'input_ids': <tf.Tensor: shape=(2, 17), dtype=int32, numpy=
array([[44232, 13354,  7797,   475, 19989,  8668, 12848,    13, 50256,
        50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256],
       [36479,    72,   447,   247,    82, 35003, 11108,   373, 40181,
         1028, 30683,   447,   247,    82,  1029,  1803,    13]],
      dtype=int32)>, 'attention_mask': <tf.Tensor: shape=(2, 17), dtype=int32, numpy=
array([[1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]], dtype=int32)>}


### Explanation of the Tokenization Output



After tokenizing our football-specific training data, we received a dictionary with two keys: `input_ids` and `attention_mask`.

- **`input_ids`**:  
  This is a TensorFlow tensor containing the numerical IDs for each token in the input sequences.  
  - **Shape:** `(2, 17)` indicates that we have 2 input texts, each padded or truncated to 17 tokens.  
  - The numbers (such as 44232, 13354, etc.) are the token IDs that correspond to words or subword units from the GPT-2 vocabulary.
  - Notice that for the first example, after a certain point (in this case, from index 8 onward), the IDs are `50256`. Since we set the pad token to the GPT-2 `eos_token`, these `50256` values represent padding tokens.

- **`attention_mask`**:  
  This tensor indicates which tokens are actual data (`1`) and which are padding (`0`).  
  - **Shape:** Also `(2, 17)`, matching the input IDs.
  - For the first text, the attention mask shows `[1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0]`, meaning that the first 8 tokens are real tokens, and the remaining 9 are padding.
  - For the second text, the attention mask is all ones, indicating that no padding was required.

This output confirms that the text data has been successfully tokenized, padded, and converted into TensorFlow tensors, ready to be fed into the model for training.


### 4. Creating a TensorFlow Dataset



**Dataset Creation:**  
We convert the tokenized output (a dictionary of tensors) into a TensorFlow dataset using `tf.data.Dataset.from_tensor_slices()`.  
For language modeling, we set the labels to be the same as the input IDs.

**Training Setup:**  
We compile the model with an optimizer, a loss function, and evaluation metrics, then fine-tune using `model.fit()`.


In [29]:
# Create a TensorFlow dataset from the tokenized encodings.
# For language modeling, we use the input_ids as both inputs and labels.
train_dataset = tf.data.Dataset.from_tensor_slices({
    "input_ids": train_encodings["input_ids"],
    "attention_mask": train_encodings["attention_mask"],
    "labels": train_encodings["input_ids"]  # For language modeling, labels are identical to input_ids
})

# Batch the dataset (using a batch size of 2)
train_dataset = train_dataset.batch(2)

# Compile the model
# For language modeling with GPT2, you might use sparse categorical crossentropy
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Fine-tune the model using model.fit()
model.fit(train_dataset, epochs=3)


Epoch 1/3
1/1 [==============================] - 10s 10s/step - loss: 10.8249 - accuracy: 0.0000e+00
Epoch 2/3
1/1 [==============================] - 0s 395ms/step - loss: 10.8249 - accuracy: 0.0000e+00
Epoch 3/3
1/1 [==============================] - 0s 148ms/step - loss: 10.8249 - accuracy: 0.0000e+00


**What Does This Mean?**

1. **Single Batch per Epoch:**  
   - The dataset we created contains only two examples, and we batch them with a batch size of 2. Thus, each epoch processes exactly one batch.

2. **Constant Loss Value (≈ 10.8249):**  
   - The loss remains constant across epochs, which indicates that the model isn’t improving on this very small dataset.  
   - The high loss value is expected in this scenario, as the model’s initial predictions are far off from the target values when using a complex model like GPT-2 for language modeling.
  
3. **Zero Accuracy:**  
   - Accuracy in language modeling (especially using metrics like sparse categorical crossentropy) is often not very meaningful.  
   - Accuracy here represents the percentage of tokens that were predicted exactly correctly, and with such a small dataset and a large vocabulary, achieving any token-level accuracy is very challenging.
  
4. **Data Limitations:**  
   - Our training dataset consists of only two sentences. This is far too small for effective fine-tuning of a model as large as GPT-2.  
   - In a real-world scenario, you would use a much larger dataset of football-specific commentary to meaningfully adjust the model's weights.

**Summary:**  
The observed output (constant loss and zero accuracy) is typical for a demonstration using an extremely small dataset. For significant improvements and meaningful training metrics, a larger dataset and possibly further hyperparameter tuning (such as adjusting the learning rate) would be necessary.



### Recap and Explanation



1. **Model and Tokenizer Loading:**  
   We loaded the TensorFlow version of GPT‑2 and ensured the tokenizer has a pad token (using eos_token if necessary).

2. **Data Preparation:**  
   A small dataset of football-specific commentary was defined and tokenized.  
   The tokenization ensures each text is truncated or padded to a consistent length (64 tokens).

3. **Dataset Creation:**  
   The tokenized data was converted into a TensorFlow dataset.  
   In our language modeling task, the input IDs serve as both inputs and labels.

4. **Training Setup and Fine-Tuning:**  
   We compiled the model using the Adam optimizer and sparse categorical crossentropy loss.  
   Finally, we fine-tuned the model using the native TensorFlow `model.fit()` method.


✅ **Why Fine-Tune GPT?**  
- GPT **already knows general language**.  
- We **only need to adjust it for football conversations**.

# **🚀 Key Takeaways: Fine-Tuning for Sports Analytics**

| Use Case                                | Pre-Trained Model                   | Fine-Tuning Needed?                      |
|-----------------------------------------|-------------------------------------|------------------------------------------|
| **Football Player Recognition** 📷⚽     | ResNet50 (ImageNet)                 | Yes (train last layers on football images) ✅ |
| **NBA Play Classification** 🎥🏀         | I3D/C3D (Sports Videos)             | Yes (train on NBA clips) ✅                |
| **Football Chatbot (NLP)** 🗣️⚽           | GPT-4, BERT                         | Yes (train on football text) ✅            |

✅ **Fine-Tuning is powerful when:**  
- **You don’t have enough data** to train from scratch.  
- **You need domain-specific adaptation** (football, NBA, medical, finance).  
- **The pre-trained model has general knowledge** that can be applied to your task.

# **🔥 Final Thoughts**

1️⃣ **Fine-Tuning = Adjusting Pre-Trained Models for New, Related Tasks.**  
2️⃣ **It saves time & computational resources.**  
3️⃣ **Used in image recognition, video analysis, NLP, and self-driving sports analysis.**  
4️⃣ **Perfect for player detection, action classification, and chatbot creation.**